File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class Trie {

    class TrieNode
    {
        HashMap<Character, TrieNode> children = new HashMap<>();
        boolean isEnd;
    }

    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode curr = root;        
        for (char c: word.toCharArray())
        {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
    }
    
    public boolean search(String word) {
        TrieNode curr = root;
        for (char c: word.toCharArray())
        {
            curr = curr.children.get(c);
            if (curr == null) return false;
        }
        
        return curr.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for (char c: prefix.toCharArray())
        {
            curr = curr.children.get(c);
            if (curr == null) return false;
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode() 

    def insert(self, word: str) -> None:
        curr = self.root
        
        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        
        curr.isEnd = True

    def search(self, word: str) -> bool:
        curr = self.root
        
        for c in word:
            curr = curr.children[c]
            if curr is None: return False
        
        return curr.isEnd

    def startsWith(self, prefix: str) -> bool:
        curr = self.root

        for c in prefix:
            curr = curr.children[c]
            if curr is None: return False

        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)